src/analyses/local_control_flow_history.cpp

```diff
@@ -0,0 +1,380 @@
+  // not be that "fair" when limiting the histories per location.
+  if(this->current_loc->location_number < other.current_loc->location_number)
+  {
+    return true;
+  }
+  else if(
+    this->current_loc->location_number > other.current_loc->location_number)
+  {
+    return false;
+  }
+  else
+  {
+    if(
+      this->control_flow_decision_history ==
+      other.control_flow_decision_history)
+    {
+      // They are equal so...
+      return false;
+    }
+    else if(other.control_flow_decision_history == nullptr)
+    {
+      // Special case so that the merged control flow history is last step's
+      // others set also means that non-merged histories are strictly preferred
+      // on the work queue
+      return true;
+    }
+    else if(this->control_flow_decision_history == nullptr)
+    {
+      // Another one, also guarding what we are about to do...
+      return false;
+    }
+    else
+    {
+      return (
+        *(this->control_flow_decision_history) <
+        *(other.control_flow_decision_history));
+    }
+  }
+  UNREACHABLE;
+}
+
+template <bool track_forward_jumps, bool track_backward_jumps>
+bool local_control_flow_historyt<track_forward_jumps, track_backward_jumps>::
+operator==(const ai_history_baset &op) const
+{
+  auto other = dynamic_cast<
+    const local_control_flow_historyt<track_forward_jumps, track_backward_jumps>
+      &>(op);
+
+  // Make the short-cutting clear
+  if(this->current_loc->location_number == other.current_loc->location_number)
+  {
+    if(
+      this->control_flow_decision_history ==
+      other.control_flow_decision_history)
+    {
+      return true;
+    }
+    else if(
+      this->control_flow_decision_history == nullptr ||
+      other.control_flow_decision_history == nullptr)
+    {
+      // Only one is null so clearly can't be equal
+      return false;
+    }
+    else
+    {
+      // Don't have unique construction so in things like step it is possible
+      // to get two distinct local_control_flow_decisiont objects that are equal
+      // in value.
+      return (
+        *(this->control_flow_decision_history) ==
+        *(other.control_flow_decision_history));
+    }
+  }
+  else
+  {
+    return false;
+  }
+  UNREACHABLE;
+}
+
+template <bool track_forward_jumps, bool track_backward_jumps>
+void local_control_flow_historyt<track_forward_jumps, track_backward_jumps>::
+  output(std::ostream &out) const
+{
+  out << "local_control_flow_history : location "
+      << current_loc->location_number;
+  lcfd_ptrt working = control_flow_decision_history;
+  while(working != nullptr)
+  {
+    out << " after " << working->branch_location->location_number;
+    if(working->branch_taken)
+    {
+      out << " taken";
+    }
+    else
+    {
+      out << " not taken";
+    }
+    working = working->previous;
+  }
+  return;
+}
+
+bool local_control_flow_decisiont::
+operator<(const local_control_flow_decisiont &op) const
```

Comment:
Looks like you intentionally want more informative orderings to be lower in the sort order? If so say so up top

Reply:
Ummm... I am not entirely sure I have an idea of what constitutes informative in this context.  Where the choice of ordering matters (both in this and in `local_control_flow_historyt<track_forward_jumps, track_backward_jumps>::
operator<`) it is commented and in one case checked by assertion (if there is a merge history it should sort last).  Specifically this includes:
```
// Lower locations first, generally this means depth first
// Branch not taken takes priority
// This prioritises complete histories over merged ones
```